### DISCUSSION TITLE
[Extending Admin API] Overriding generated admin api  (partially)

### DISCUSSION THREAD

**Discussion by TVenuMadhav** [Kong/kong#7497] [Help]
In lua plugin development, we have an option to define our own custom schema(s) and while doing that we also have an option to generate admin api  by setting the `generate_admin_api = true` in schema config (daos.lua)

Now let's say the admin api generated are 'list, create, retrieve, update & delete' on an entity and the usage of these admin api are as follows,

Examples:

List => GET /abc/
Retrieve => GET /abc/:id/
Create => POST /abc/
Update => PUT /abc/:id/
Delete => DELETE /abc/:id/

Now suppose if I want to override the DELETE admin api (to include some additional functionality for example), i would generally need to define the custom admin api implementation in api.lua file which goes something like,

```lua
['/abc/:id'] = {
    DELETE = function(self)
    ...
    end
}
```

# What I expect
Now that we have overridden the DELETE admin api only, I expect all the other methods (Retrieve, update) to work as generated by default.  

# What I am experiencing

😕  But that is not the case, instead it looks like all the admin api methods that involve the uri signature `/abc/:id` stand overridden.

Now only DELETE works are I have defined but GET '/abc/:id' & PUT '/abc/:id/' are undefined. I get method not allowed error on calling the retrieve and update admin apis!

Hope this makes sense! 

I would love to get some help regarding this as the documentation isn't very clear on this [[Ref](https://docs.konghq.com/gateway-oss/2.4.x/plugin-development/admin-api/)] 🙏🏻

**Comment by TVenuMadhav** [comment]
hoping for an acknowledgement / possible solution

**Reply by Tieske** [reply]
this does what you want no? https://github.com/Kong/kong/blob/master/kong/plugins/acl/api.lua

**Reply by TVenuMadhav** [reply]
Hi @Tieske,

Thank you for your response. 

From what I understand, `parent` is passed as an argument in the customised Admin api (overridden version of generated admin api) - Using comments at [[Ref](https://github.com/Kong/kong/blob/911385466571e4f72999fa43419daafae08159d0/kong/api/init.lua#L45)]

So to reproduce the same, I updated my custom admin api to 
```
['/abc/:id'] = {
    GET = function(self, _, _, parent)
       return parent()
    ...
    end,
    PUT = function(self, _, _, parent)
       return parent()
    ...
    end,
    DELETE = function(self)
    -- my custom delete implementation 
    ...
    end
    
}
```
However, on making GET and PUT calls, I get,
```
2021/06/30 10:27:18 [error] 22#0: *934 [lua] api_helpers.lua:422: handle_error(): /usr/local/share/lua/5.1/lapis/application.lua:401: /usr/local/share/lua/5.1/kong/plugins/acme/api.lua:79: attempt to call local 'parent' (a nil value)
```

I checked how things are usually overridden at [[Ref#1](https://github.com/Kong/kong/blob/911385466571e4f72999fa43419daafae08159d0/kong/api/routes/certificates.lua#L109)]

Is there something crucial  I am missing ?

**Reply by Tieske** [reply]
@bungle iirc you implemented the auto-generated endpoints. Do you know if and how they can be overridden?

**Comment by kikito** [comment]
@TVenuMadhav part of the trouble you are experiencing is due to the fact that the Admin API "hooks" have not been properly formalized yet. As a result, the docs are not up to date so it is very difficult to understand how things work.

Is your plugin code available somewhere? If it isn't possible for you to share it, could you include the parts around this?

```
['/abc/:id'] = {
    DELETE = function(self)
    ...
    end
}
```

The code surrounding that will define when/how that route is added.

**Comment by bungle** [comment]
You cannot just use anything like `:id`. You need to specify it similarly as our generator does. Likely `/abc/:abc` (use the dao name here).

See: https://github.com/Kong/kong/blob/master/kong/api/endpoints.lua#L413-L414
